### DISCUSSION TITLE
How to secure keycloak client id and client secret when integrating an Angular application with keycloak?

### DISCUSSION THREAD

**Discussion by vandanakulkarni05** [keycloak/keycloak#30810] [Q&A] [answered]
I have an angular web application that is deployed as a docker container. We are currently keeping the Keycloak client id and client secret in environment variables, however, I do not want to keep this in environment variables in the future. Is there any better and secure approach to do this?

**Comment by jonkoops** [comment]
Use a standard OpenID Connect Authorization Code Flow with PKCE enabled, that is the most secure way to do this. You cannot store a client secret in a web application, as it can be read in the browser and stolen to make malicious requests.

**Reply by vandanakulkarni05** [reply]
Thank you for the response. I wanted to understand what is the difference in Implicit flow and PKCE enabled flow if you could help?

**Reply by jonkoops** [reply, marked as answer]
PKCE is just an extension of the Authorization Code flow, the Implicit Flow returns a token directly, but will not provide you with a refresh token, requiring the user to re-authenticate when it expires. I would recommend you read [our documentation](https://www.keycloak.org/docs/latest/securing_apps/index.html#supported-grant-types) on this subject.

**Reply by vandanakulkarni05** [reply]
Thanks for all the help @jonkoops

**Reply by jonkoops** [reply]
No problem!
